org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/ParallelBuilderParticipant.java

```diff
@@ -230,7 +230,7 @@ protected void doBuild(
 			List<SimpleEntry<URI, Throwable>> exceptions = Lists.newArrayList();
 			boolean interrupted = false;
 			try {
-				while (!requestQueue.isEmpty() || !generatorResult.isDone()) {
+				while (!generatorResult.isDone() || !requestQueue.isEmpty()) {
```

Comment:
my tests over the weekend with the reproducer main as well as with my customers production tests look good so far.